worker/changestream/metrics.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package changestream
+
+import "github.com/prometheus/client_golang/prometheus"
+
+const (
+	changestreamMetricsNamespace   = "juju"
+	changestreamSubsystemNamespace = "db"
+)
+
+var (
+	labelNames = []string{"namespace"}
+)
+
+// Collector defines a prometheus collector for the dbaccessor.
+type Collector struct {
+	Namespace string
+	// Stream metrics.
+	WatermarkInserts       *prometheus.CounterVec
+	WatermarkRetries       *prometheus.CounterVec
+	ChangesRequestDuration *prometheus.HistogramVec
+	ChangesCount           *prometheus.HistogramVec
+	// EventMultiplexer metrics.
+	Subscriptions    prometheus.Gauge
+	DispatchDuration *prometheus.HistogramVec
+	DispatchErrors   *prometheus.CounterVec
+}
+
+func (c *Collector) WatermarkInsertsInc() {
+	c.WatermarkInserts.WithLabelValues(c.Namespace).Inc()
+}
+
+func (c *Collector) WatermarkRetriesInc() {
+	c.WatermarkRetries.WithLabelValues(c.Namespace).Inc()
+}
+
+func (c *Collector) ChangesRequestDurationObserve(val float64) {
+	c.ChangesRequestDuration.WithLabelValues(c.Namespace).Observe(val)
+}
+
+func (c *Collector) ChangesCountObserve(val int) {
+	c.ChangesCount.WithLabelValues(c.Namespace).Observe(float64(val))
+}
+
+func (c *Collector) SubscriptionsInc() {
+	c.Subscriptions.Inc()
+}
+
+func (c *Collector) SubscriptionsDec() {
+	c.Subscriptions.Dec()
+}
+
+func (c *Collector) SubscriptionsClear() {
+	c.Subscriptions.Set(0)
+}
+
+func (c *Collector) DispatchDurationObserve(val float64) {
+	c.DispatchDuration.WithLabelValues(c.Namespace).Observe(val)
+}
+
+func (c *Collector) DispatchErrorsInc() {
+	c.DispatchErrors.WithLabelValues(c.Namespace).Inc()
+}
+
+func (c *Collector) ForNamespace(namespace string) *Collector {
+	child := *c
+	child.Namespace = namespace
+
+	return &child
+}
```

Comment:
Just thinking about this now, the collector should just be of the following type, I know it's annoying, but it's a lot cleaner. 

```
type Metrics interface {
    ForNamespace(string) Collector
    Describe(ch chan<- *prometheus.Desc)
    Collect(ch chan<- prometheus.Metric)
}

type Collector interface {
      ...
}
``` 

See https://github.com/juju/juju/blob/94735455727231499b1e5a2bac07b97cb6575b0e/agent/engine/metrics.go#L68-L73

Reply:
done